### ISSUE TITLE
GPU copying slow from loss functions

### ISSUE THREAD

**Issue by TiRune** [pytorch/ignite#240]
Hey all,

I did some profiling of the ignite code:
![screen shot 2018-08-17 at 17 02 40](https://user-images.githubusercontent.com/1436356/44273523-0036d400-a240-11e8-9db4-b19a50a54121.png)

Looks like in my training runs, around 20% of the time gets used by the loss functions. Specifically this happens for the .item() calls in pytorch. The metrics are copied from cuda to gpu every iteration, which is very slow. This happens in both TopKCategoricalAccuracy and Loss
Perhaps we should replace the .item() call, and put it in compute, so we gather the statistics on the gpu and keep them there until they are used.

**Comment by vfdev-5** [comment]
@TiRune thanks for the feedback, do you mind to share the code to reproduce the profiling please ?
As I understand correctly you run the training + online metrics computation.

> Perhaps we should replace the .item() call, and put it in compute, so we gather the statistics on the gpu and keep them there until they are used.

Yes, we need to take a closer look at this...

**Comment by TiRune** [comment]
Can't share the code unfortunately, company rules.

Yes I'm doing training, and try to gather statistics of the training loss every few iterations.
I found out that this was causing slowdowns when reading: https://www.sagivtech.com/2017/09/19/optimizing-pytorch-training-code/

Who suggested the methods for profiling, and also noted that (Cuda copies are expensive.

> It turned out that a lot of our cuda copies were for batch statistics: the loss, accuracy, and other data.
> Instead of displaying the loss and other metrics for every batch, aggregate them on the GPU and copy them to the CPU for display at the end of every epoch.)

update gets called every training iteration, potentially causing this slowdown.

**Comment by jasonkriss** [comment]
Interesting. Thank you for this @TiRune! This seems very reasonable to me. The only real concern I can see is that GPU memory is generally a much more limited resource. Seems to me like this outweighs that concern though.

**Comment by alykhantejani** [comment]
@TiRune thanks for the bug report, this looks interesting. It would be great if you could confirm if this happens on master. If so, we should fix this asap

**Comment by TiRune** [comment]
Looking into it on my side as well. Won't be done today because of some hectic deadlines, but I'll keep you posted :)
Happily using ignite btw, so great work :)

**Comment by alykhantejani** [comment]
Thanks @TiRune - looking forward to your response

**Comment by TiRune** [comment]
The new master branch of ignite definitely reduced the overhead time significantly; so that is great.

I also figured out that if I remove the loss function calculation all together; the same overhead is now taken by prepare_batch... So I can only conclude that the cprofile code is wrong somehow; and it's confusing timings completely. 
It's very likely the overhead is caused by prepare_batch instead; thus I think it's safe to close this ticket.

**Comment by vfdev-5** [comment]
@TiRune thanks for the feedback! Are you obtaining something similar as in my [previous message](https://github.com/pytorch/ignite/issues/240#issuecomment-414002722) ?

Anyway glad that we have a similar conclusion on metric computation.

**Comment by alykhantejani** [comment]
@TiRune that's good to know. We will be releasing a new update soon to conda and pip so you wont have to build from master (probably in 2 weeks)

**Comment by TiRune** [comment]
@vfdev-5 Interestingly for me; although I had multiple losses logged, it only logs a bunch of time on 1 loss. My cProfile is not nearly as fine-grained as yours. No idea why this is though :/. 

See:
<img width="981" alt="screen shot 2018-08-23 at 15 03 13" src="https://user-images.githubusercontent.com/1436356/44526932-b6d60100-a6e5-11e8-852b-2bd7d72fae02.png">
Think Cprofile is a bit broken for multi-gpu models or because of the multi-threaded data-loaders

**Comment by vfdev-5** [comment]
@TiRune Okay, I see. In my case I tested on a single GPU, python 3.5, snakeviz version 1.0.0

**Comment by TiRune** [comment]
Alright, here's what I think now happens:
GPU calls are async; function waits in .item() when it's called (in this case top_k_categorical.update is called first of all the loss functions), thus it's stuck there until the cuda is synced again. Thus it looks like .item() is causing slowdowns; but in actuality the code is still running the forward pass.

Case closed it seems; cProfile can't really be trusted with multiple workers and multi-gpu :)

**Comment by vfdev-5** [comment]
@TiRune maybe you've already seen [`torch.utils.bottleneck`](https://pytorch.org/docs/stable/bottleneck.html), in any case it seems they say something similar.
